libcontainer/factory_linux.go

```diff
@@ -383,24 +383,22 @@ func (l *LinuxFactory) StartInitialization() (err error) {
 
 	// Get mount files (O_PATH).
 	// We assume in other parts that mountFiles is nil when there are no fds to mount.
-	var mountFiles []*os.File
+	var mountFiles []*int
```

Comment:
Hmm, I'd like to avoid that. Being a pointer, the 0 value is nil. Being an int, the 0 value is 0 and therefore we need to specify it as -1 in all cases. Furthermore, if one case is added in the future and forgets to specify -1 as the fd, it will use 0 and have a really weird behavior to debug.

Reply:
Oh, wait, you mean to change it only here? Let me check... Sorry, I thought the context was different. Let me re-check...

Reply:
Yeap, changed this to []int instead.

While doing this, I had to make the [this line][link] just be `... := []int{}` and append to the list. We need to insert just the numbers that we want, otherwise if we use `make` with a longer length (for non bind-mounts we don't want to insert anything) the 0 value (0) will be used that is another, and usually valid, fd.


[link]: https://github.com/opencontainers/runc/pull/2576/files#diff-08395501389b35b3ff26335eeaa81c2e113672e4feb4f500a3231e2f06128268R390

Reply:
While I was there, I also made the shouldSendMountSources check, [also check for bind-mounts][link]. If there are no bindmounts, we won't send.

[link]: https://github.com/opencontainers/runc/pull/2576/files#diff-1cdede3fe918430ca0b7e4056abc49abca8a2925c6770ec1951befa6091f7a94R525-R531

Reply:
No, this change will probably break this part: https://github.com/opencontainers/runc/pull/2576/files#diff-e0932af06cb46065b2ef5fc89bc2bd9b880c9578e01b89ef4ecaeffb78a8bef0R76-R78

I think we need to keep the []*int, as we won't know the fd for which mount should be used otherwise.

Grr, going to revert that and add a test with another bindmount that actually exercise this test next week. Sorry :(

Reply:
Ok, so this was not broken by chance, but I fixed it and added a few barriers so we don't break it.

The thing is, indeed the code expects this slice to have the same len as config.Mounts. Then, we parse the string created here and assign an fd (if any) to the proper mount. However, when we prepare the netlink-alike message, we only append the need to ask for sources to bind mounts (see [the code here][link]), so that was "hiding" the issue and the proper mounts were seen, with the proper content.

However, that is super inconsistent and asking for issues, so I returned back the code to use *int. This way, we use "nil" for fds that are not present and that is all.

However, as that is quite an important thing to note in the code for future readers, I made it way more explicit and added a few sanity checks with clear errors. I added:
 * shouldSendMountSources() now checks for bind-mounts too. If there are no bind-mounts, it returns false. Before this, it would return true but if there are no bind mounts no fds are opened and therefore the _LIBCONTAINER_MOUNT_FDS was just a collection of ";"
 * Added a check in rootfs_linux.go: `if len(mountFds) > 0 && len(mountFds) != len(config.Mounts)` --> error out. The code was not doing this but just checking [if the size was not exceeded][link-not-size], when it should always be the same size.
 * However, it was not the same size due to the say the _LIBCONTAINER_MOUNT_FDS str was constructed and parsed. The str was constructed as "number;" and then append (either using strings.Builder or in the old version with plain strings and append). Therefore, it always terminated in ";" and when parsed in container_linux.go with `strings.Split(str, ";")`, this creates always an empty element for the element after the last ";". IOW, the size was always one more.
 * For that reason, I removed the usage of strings.Builder and create just create a `[]string` with the needed strings, and then join it with `strings.Join(slice, ";")`. This creates a string that we can parse with `strings.Split()` and has the expected size.
 * I've also added some more comments in the code to make it very clear what the size of this slice should be and why.
 * Note that this will indeed add some more allocations, but the code reads so clean that I think this is better than creating the string with the ending ";" and having a big comment to explain why, when we parse it with `strings.Split(..., ";")` we use `len(...) - 1` and ignore the last element. Let me know if you prefer otherwise.

As that bug is not really hit in an e2e test (as it is hidden due to the netlink-alike part that adds the request for sources), I couldn't add a test for that. However, we do have a nice check in the code that runc fails with a clear error if the mounts and mountFds slice are not of the same size (item 2 of the previous list is what I added).

I'm pushing now with these improvements.

[link]: https://github.com/opencontainers/runc/pull/2576/files#diff-1cdede3fe918430ca0b7e4056abc49abca8a2925c6770ec1951befa6091f7a94R2158-R2172

[link-not-size]: https://github.com/opencontainers/runc/pull/2576/files#diff-e0932af06cb46065b2ef5fc89bc2bd9b880c9578e01b89ef4ecaeffb78a8bef0R76-R78

Reply:
Re: last `;`, you could just remove this after the loop, with a simple explanation. But I agree that code readability matters, and maybe strings.Join() is not that bad.

Reply:
Cool. The ""issue"" with removing the last `;` is that if we use `strings.Builder`, we can only append. So, we would need to convert to string and then remove it.

It is completely possible too, but not as clean as I'd like it to be.